logstash-core/src/main/java/org/logstash/common/io/DeadLetterQueueReader.java

```diff
@@ -214,24 +242,44 @@ private Path nextExistingSegmentFile(Path currentSegmentPath) {
 
     public void setCurrentReaderAndPosition(Path segmentPath, long position) throws IOException {
         // If the provided segment Path exist, then set the reader to start from the supplied position
-        if (Files.exists(segmentPath)) {
-            currentReader = new RecordIOReader(segmentPath);
+        Optional<RecordIOReader> optReader = openSegmentReader(segmentPath);
```

Comment:
Is there a reason why we're not removing `segmentPath` here, like we are with the other invocations of `openSegmentReader`? If not, I wonder if we put the `segments.remove(segmentPath)` in the `openSegmentReader` method to ensure that we clean up afterwards, and it might help clean up some of the calling code

Reply:
Effectively you are right, we can clean up the segment under opening directly inside `openSegmentReader` method.

Reply:
Suggestion implemented with commit 358f31dd6b5908e4ffcfcaf18cd1e532097874cc